src/metabase/driver/sql/query_processor.clj

```diff
@@ -630,18 +696,36 @@
 ;; also, we want to gracefully handle situations where the column is ZERO and just swap it out with NULL instead, so
 ;; we don't get divide by zero errors. SQL DBs always return NULL when dividing by NULL (AFAIK)
 
+(defn- safe-denominator
+  "Make sure we're not trying to divide by zero."
+  [denominator]
+  (cond
+    ;; try not to generate hairy nonsense like `CASE WHERE 7.0 = 0 THEN NULL ELSE 7.0` if we're dealing with number
+    ;; literals and can determine this stuff ahead of time.
+    (and (number? denominator)
+         (zero? denominator))
+    nil
+
+    (number? denominator)
+    (inline-num denominator)
+
+    (inline? denominator)
+    (recur (second denominator))
+
+    :else
+    (hx/call :case
+             (hx/call := denominator (inline-num 0)) nil
+             :else                                   denominator)))
+
 (defmethod ->honeysql [:sql :/]
-  [driver [_ & args]]
-  (let [[numerator & denominators] (for [arg args]
-                                     (->honeysql driver (if (integer? arg)
-                                                          (double arg)
-                                                          arg)))]
+  [driver [_ & mbql-exprs]]
+  (let [[numerator & denominators] (for [mbql-expr mbql-exprs]
+                                     (->honeysql driver (if (integer? mbql-expr)
+                                                          (double mbql-expr)
+                                                          mbql-expr)))]
     (apply hx/call :/
            (->float driver numerator)
-           (for [denominator denominators]
-             (hx/call :case
-                      (hx/call := denominator 0) nil
-                      :else                        denominator)))))
+           (map safe-denominator denominators))))
```

Comment:
Good point. Not really anything to do with the changes here since that's old code but I'll look into fixing it

Reply:
Ok, not sure we want to remove that call to `->float` entirely, since I'm not sure which databases have a separate `double` type in addition to `float`, but I will tweak it so it avoids a cast if it's already a `java.lang.Double`

Reply:
Not fixed 100% but I tweaked the code to optimize out the `CAST(x to FLOAT)` call when we already have a number literal. Don't want to change this to `->double` or something because I think some DB types only have `float` but it's 64-bit. Not sure about this but I don't want to fix things that aren't broken in this PR